### ISSUE TITLE
All xtrsm tests fail

### ISSUE THREAD

**Issue by Ulfgard** [CNugteren/CLBlast#152]
This issue reproduces on the latest release as well as develop. I build the library and tests and run the tests. All tests pass except xtrsm, which fails almost completely. This is on a cpu device.

I am attaching the error log as well as the output of clinfo

[xtrsm_results.txt](https://github.com/CNugteren/CLBlast/files/995479/xtrsm_results.txt)
[clinfo.txt](https://github.com/CNugteren/CLBlast/files/995480/clinfo.txt)

should i try tuning?

**Comment by CNugteren** [comment]
Thanks for reporting! Yes, actually this is a known issue, I should have wrote that down somewhere. I am still not sure whether it is an actual problem in the TRSM routine or in the test set-up. This is on my list to fix soon, I'll keep you up to date.

**Comment by CNugteren** [comment]
I prioritised this and I made some fixes. On my machine the TRSM tests now pass. Could you try again with the latest version of `master`?

**Comment by Ulfgard** [comment]
Sorry, this still fails for me.
[xtrsm_log.txt](https://github.com/CNugteren/CLBlast/files/998975/xtrsm_log.txt)

**Comment by CNugteren** [comment]
OK, but it seems to have improved ;-) First you reported e.g. for STRSM:
```
   140 test(s) passed
   208 test(s) skipped
   164 test(s) failed
```
and now for STRSM:
```
   190 test(s) passed
   208 test(s) skipped
   114 test(s) failed
```
So the fixes indeed fixed something, but not everything. I just run the tests on some other GPU devices which showed failures before, but now they all pass. I then tested on an Intel CPU on Linux and indeed I see a similar pattern of failures as well. I'll investigate it!

**Comment by Ulfgard** [comment]
Hi,

It is most likely that the issue is a missing memory barrier in some kernel. I have not much time today to check (and am not familiar with the code base) but i would check whether the inverse of the diagonal blocks assumes that reads/writes are in lock-step. The rest looks reasonable to me.

Best,
Oswin

________________________________

**Comment by CNugteren** [comment]
Thanks for your feedback :+1: I think you are right: I added a missing synchronisation barrier in the invert-diagonal-blocks kernel and now tests pass on my CPU system as well. Can you verify what happens with on your system with the latest version of `master`? Thanks!

**Comment by Ulfgard** [comment]
Thanks! Works for me!
________________________________

**Comment by Ulfgard** [comment]
Please do not close yet. I have trouble running Trsm correctly in my framework. The case: 
AX=B with row-major lower triangular works fine. But row-major upper triangular does not. I am comparing with your tests.

**Comment by Ulfgard** [comment]
okay. I could not understand your test framework. However, i do not get a wrong result but an errorcode != 0. So if Trinangular is kLower, everything works fine. If it is kUpper, i get error 

kInsufficientMemoryB       = -1010

This is weird, because i do not see what the triangular should change in the memory requirements, the results should be the same. I also checked that it is independent of the layout (checking both layouts of B)

**Comment by Ulfgard** [comment]
Here is the setup:
layout: column-major
upper=true
Side=left
Diagonal: kNonUnit(does not matter)

m=139
n=238

An exception is thrown by DoGemm in line 159 of xtrsm.cpp when testing
TestMatrixB(b_one, b_two, b_buffer, b_offset, b_ld); (line 101 xgemm.cpp)

unfortunately even in debug mode all numbers are optimized out, so i can not check what is going wrong.

**Comment by CNugteren** [comment]
You are right, there is still another issue. I'll extend the tests to cover that particular case and see if I can fix it. I'll keep you up-to-date.

**Comment by CNugteren** [comment]
OK, I've fixed it now in [#8400ee3](https://github.com/CNugteren/CLBlast/commit/8400ee3a097952a49371973780b47fcbf63e9a5f). Sorry for having to encounter all these bugs... TRSM was really still an experimental routine, not well tested as you can see. Hopefully it does now what it is supposed to?

**Comment by Ulfgard** [comment]
It is working now! Thanks for your support. 

I now have added CLBlast as backend for Shark-ML/Remora
